traits/has_traits.py

```diff
@@ -396,6 +427,13 @@ def update_traits_class_dict(class_name, bases, class_dict):
     # Move all trait definitions from the class dictionary to the
     # appropriate trait class dictionaries:
     for name, value in list(class_dict.items()):
+
+        # Warn name collisions.
+        if check_name_collision and name in public_method_names:
```

Comment:
I'm in favor of reserving names starting with `traits` or `_traits`. But do we want to do that in addition to or instead of what's being done in this PR ?

Reply:
What if `HasTraits` introduces a new public method that does not being with `traits` ?

Reply:
> What if `HasTraits` introduces a new public method that does not being with `traits` ?

So the idea of #705 is that Traits would promise not to introduce such new methods: it would use names beginning with `trait` for new public methods, and names beginning with `_trait` for new private methods.

That's exactly the problem that #705 aims to solve: every time we introduce a new method on `HasTraits` we risk conflicting with user-defined trait names. But if we only ever introduce new methods in the `trait*` / `_trait*` namespace, and we tell _users_ of Traits not to name their own traits that way, we avoid that conflict.